### PR TITLE
Add deferred toggle for ChannelSelectionOverlay if no channels are joined

### LINKED ISSUE #5161: Channel selector will not automatically open if there are no channels joined
This logic was deliberately removed in https://github.com/ppy/osu/pull/5156 until we can fix it.

This needs to account for the fact that we may be still loading in, and a simple check for JoinedChannels is not enough as it might not have been populated. In the case we are still loading, the selector should not be toggled until we can confirm there are no channels joined. 

This can potentially be done using a check for whether or not there are anything in `AvailableChannels`.

### PR DESCRIPTION
**Pull request by nyquillerium**
* Fixes https://github.com/ppy/osu/issues/5161
* [x] Depends on https://github.com/ppy/osu/pull/5156

This makes it so we don't show an empty `ChannelSelectionOverlay` when chat is toggled, but toggles it whenever available channels are loaded in.

### CODE REVIEW

File: osu.Game/Overlays/ChatOverlay.cs [hunk 1]
```diff
@@ -358,15 +355,31 @@ protected override void PopIn()
             this.FadeIn(transition_length, Easing.OutQuint);
 
             textbox.HoldFocus = true;
+
+            // Don't show an empty channel selector if no channels are loaded yet.
+            if (!channelManager.AvailableChannels.Any())
+                channelManager.AvailableChannels.ItemsAdded += toggleContinuation;
+            else if (!channelManager.JoinedChannels.Any())
```
Comment (peppy) [inline osu.Game/Overlays/ChatOverlay.cs]: duplicated code

File: osu.Game/Overlays/ChatOverlay.cs [hunk 2]
```diff
@@ -358,15 +355,31 @@ protected override void PopIn()
...
+                ChannelSelectionOverlay.Show();
+
             base.PopIn();
         }
 
+        private void toggleContinuation(IEnumerable<Channel> channels)
+        {
+            if (!channelManager.JoinedChannels.Any() && State.Value == Visibility.Visible)
```
Comment (peppy) [inline osu.Game/Overlays/ChatOverlay.cs]: state visible is always true here, right..?

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
99% sure this will fail if the user is joined to the second available channel.

**Comment by nyquillerium** [comment]
I see. Will check.

File: osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs [hunk 3]
```diff
@@ -23,17 +22,14 @@ public class TestSceneChatOverlay : ManualInputManagerTestScene
     {
         public override IReadOnlyList<Type> RequiredTypes => new[]
         {
-            typeof(ChatLine),
```
Comment (peppy) [inline osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs]: why?

File: osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs [hunk 4]
```diff
@@ -43,15 +39,9 @@ public void Setup()
...
-                ChannelManagerContainer container;
-
-                Child = container = new ChannelManagerContainer(new List<Channel> { channel1, channel2 })
-                {
-                    RelativeSizeAxes = Axes.Both,
-                };
-
-                chatOverlay = container.ChatOverlay;
-                channelManager = container.ChannelManager;
+                Child = channelManagerContainer = new ChannelManagerContainer(new List<Channel> { channel1, channel2 }) { RelativeSizeAxes = Axes.Both, };
```
Comment (peppy) [inline osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs]: why?

File: osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs [hunk 5]
```diff
@@ -84,18 +74,71 @@ public void TestCloseChannelWhileSelectorClosed()
         {
             AddStep("Join channel 1", () => channelManager.JoinChannel(channel1));
             AddStep("Join channel 2", () => channelManager.JoinChannel(channel2));
-
```
Comment (peppy) [inline osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs]: add back

File: osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs [hunk 6]
```diff
@@ -109,19 +152,25 @@ private class ChannelManagerContainer : Container
...
             private readonly List<Channel> channels;
 
             public ChannelManagerContainer(List<Channel> channels)
             {
                 this.channels = channels;
             }
 
+            public void ClearAvailable() => AvailableChannels.Clear();
+
+            public void AddRange(List<Channel> channels) => AvailableChannels.AddRange(channels);
```
Comment (peppy) [inline osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs]: remove

File: osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs [hunk 7]
```diff
@@ -109,19 +152,25 @@ private class ChannelManagerContainer : Container
...
+            public BindableList<Channel> AvailableChannels => (BindableList<Channel>)ChannelManager.AvailableChannels;
+
             private readonly List<Channel> channels;
 
             public ChannelManagerContainer(List<Channel> channels)
             {
                 this.channels = channels;
             }
 
+            public void ClearAvailable() => AvailableChannels.Clear();
```
Comment (peppy) [inline osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs]: remove

File: osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs [hunk 8]
```diff
@@ -109,19 +152,25 @@ private class ChannelManagerContainer : Container
             [Cached]
             public ChannelManager ChannelManager { get; } = new ChannelManager();
 
+            public BindableList<Channel> AvailableChannels => (BindableList<Channel>)ChannelManager.AvailableChannels;
```
Comment (peppy) [inline osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs]: remove

File: osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs [hunk 9]
```diff
@@ -84,18 +74,71 @@ public void TestCloseChannelWhileSelectorClosed()
...
+        [Test]
+        public void TestShowWhileLoadingWithJoined()
+        {
+            AddStep("Join channel 2", () => channelManager.JoinChannel(channel2));
+            setupUnloadedChannelsTest();
+            fakeInitializeChat();
+            AddAssert("Selector is still closed", () => chatOverlay.SelectionOverlayState == Visibility.Hidden);
+        }
+
+        private void fakeInitializeChat()
```
Comment (peppy) [inline osu.Game.Tests/Visual/Online/TestSceneChatOverlay.cs]: no need to call it "fake". name it to explain what it's doing
also share the code with the main SetUp method, which is basically doing the same thing anyway.

**Review by peppy — CHANGES_REQUESTED** [review]

**Comment by peppy** [comment]
Going to close this for now as it doesn't look like a clean solution to the issue (and the underlying code has changed to a point that resolving the conflicts would likely not be worthwhile).

**Pull request closed without merging**